### ISSUE TITLE
[MGPG-41] Passphrase revealed when backspacing at prompt

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-gpg-plugin#216]
**[Tony Trinh](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tony19)** opened **[MGPG-41](https://issues.apache.org/jira/browse/MGPG-41?redirect=false)** and commented

At the "GPG Passphrase" prompt, if I hit the backspace key during the entry, the passphrase is printed in cleartext with one less character. For example:

```
GPG Passphrase: ******************^R
mysecretpasswor*^R
mysecretpasswo*^R
mysecretpassw*^R
mysecretpass*^R
mysecretpas*^R
mysecretpa*^R
mysecretp*^R
mysecret*^R
mysecre*^R
mysecr*^R
mysec*^R
myse*^R
mys*^R
my*^R
m*^R
*^R
*
```

This can be fixed by replacing the `MaskingThread` with Java 6's built-in password prompt (as the [code comment](http://grepcode.com/file/repository.jboss.org/maven2/org.apache.maven.plugins/maven-gpg-plugin/1.0-alpha-4/org/apache/maven/plugin/gpg/GpgSigner.java#217) had suggested to do):

```java
Console console = System.console();
if ( console != null )
{
    pass = new String( console.readPassword( "GPG Passphrase:  " ) );
}
```

---

**Affects:** 1.4

1 votes, 3 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I second that. One could check whether we are running 1.6 and use System.Console otherwise the thread can still be used.

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

r1570549

Fixed for JVM 1.6+ using reflection. Bug will still be present on JVM 1.5 fallback code path
